packages/core/src/Site/index.js

```diff
@@ -960,61 +991,108 @@ class Site {
     });
   }
 
+  stopOngoingBuilds() {
+    this.pageGenerationContext.stopGenerationTimeThreshold = new Date();
+  }
+
   /**
    * Runs the supplied page generation tasks according to the specified mode of each task.
    * A page generation task can be a sequential generation or an asynchronous generation.
    * @param {Array<object>} pageGenerationTasks Array of page generation tasks
-   * @return {Promise<void>} A Promise that resolves once all pages have generated
+   * @returns {Promise<boolean>} A Promise that resolves to a boolean which indicates whether the generation
+   * ran to completion
    */
   async runPageGenerationTasks(pageGenerationTasks) {
     const pagesCount = pageGenerationTasks.reduce((acc, task) => acc + task.pages.length, 0);
     const progressBar = new ProgressBar(`[:bar] :current / ${pagesCount} pages built`, { total: pagesCount });
     progressBar.render();
 
+    const startTime = new Date();
+    let isCompleted = true;
     await utils.sequentialAsyncForEach(pageGenerationTasks, async (task) => {
+      if (startTime < this.pageGenerationContext.stopGenerationTimeThreshold) {
+        logger.info('Page generation stopped');
+        logger.debug('Page generation stopped at generation task queue');
+        isCompleted = false;
+        return;
+      }
+
       if (task.mode === 'sequential') {
-        await this.generatePagesSequential(task.pages, progressBar);
+        isCompleted = await this.generatePagesSequential(task.pages, progressBar);
       } else {
-        await this.generatePagesAsyncThrottled(task.pages, progressBar);
+        isCompleted = await this.generatePagesAsyncThrottled(task.pages, progressBar);
       }
     });
+    return isCompleted;
   }
 
   /**
    * Generate pages sequentially. That is, the pages are generated
    * one-by-one in order.
    * @param {Array<Page>} pages Pages to be generated
    * @param {ProgressBar} progressBar Progress bar of the overall generation process
-   * @returns {Promise<void>} A Promise that resolves once all pages have been generated
+   * @returns {Promise<boolean>} A Promise that resolves to a boolean which indicates whether the generation
+   * ran to completion
    */
   async generatePagesSequential(pages, progressBar) {
+    const startTime = new Date();
+    let isCompleted = true;
     await utils.sequentialAsyncForEach(pages, async (page) => {
+      if (startTime < this.pageGenerationContext.stopGenerationTimeThreshold) {
```

Comment:
Hmm now that I think of it, I was putting this as a safeguard for when sequential generation is in play (e.g. regenerating affected pages). I guess this can catch the case when "regenerating affected pages" routine is running but user edits another file so it stops the old routine and do it again. 

So consequently the "stop generating" signal is not just for the background build, but for any build that happens when the fs events got triggered. It's a bit broader in scope than just background build process.

Do you think I should keep this?

Reply:
Ah I see, if that's the case then I think it's fine to keep this!